Define the `<selectedcontent>` element

josepharhar opened the pull request:
The `<selectedcontent>` element is part of the customizable `<select>` proposal: https://github.com/whatwg/html/issues/9799

It allows authors to declaratively clone the contents of the currently selected `<option>` of a `<select>` and style it independently for use in a base appearance `<select>`'s button.

Some details on the cloning behavior, which were decided on in order to always make sure that the selectedcontent element is up to date and in order to reduce developer confusion:
- Cloning is done with an algorithm similar to the Node.cloneNode() function.
- Cloning is done synchronously at these times:
  - When an `</option>` end tag is parsed during HTML parsing.
  - A selectedcontent element is inserted, during its post-connection steps.
  - Immediately before input and change events are fired on a select element.
  - When the select.selectedIndex IDL is assigned to.

Fixes https://github.com/whatwg/html/issues/10520

The selectedcontent element has been discussed generally here: https://github.com/w3c/csswg-drafts/issues/10242

html-aria PR: https://github.com/w3c/html-aria/pull/528
html-aam PR: https://github.com/w3c/aria/pull/2344 <!-- old: https://github.com/w3c/html-aam/pull/566-->

<!--
Thank you for contributing to the HTML Standard! Please describe the change you are making and complete the checklist below if your change is not editorial.
When editing this comment after the PR is created, check that PR-Preview doesn't overwrite your changes.
If you think your PR is ready to land, please double-check that the build is passing and the checklist is complete before pinging.
-->

- [x] At least two implementers are interested (and none opposed):
   * Chrome
   * Mozilla: https://github.com/mozilla/standards-positions/issues/1060
   * WebKit: https://github.com/WebKit/standards-positions/issues/386
- [x] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * https://wpt.fyi/results/html/semantics/forms/the-select-element/customizable-select/selectedcontent.tentative.html
- [x] [Implementation bugs](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chromium: [Already implemented](https://issues.chromium.org/issues/40146374)
   * Gecko: https://bugzilla.mozilla.org/show_bug.cgi?id=1944403
   * WebKit: https://bugs.webkit.org/show_bug.cgi?id=286642
- [x] [MDN issue](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) is filed: https://github.com/mdn/content/issues/37857
- [x] The top of this comment includes a [clear commit message](https://github.com/whatwg/meta/blob/main/COMMITTING.md) to use. <!-- If you created this PR from a single commit, Github copied its message. Otherwise, you need to add a commit message yourself. -->

(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/html/10633/form-control-infrastructure.html" title="Last updated on May 5, 2025, 2:02 PM UTC (bcd9669)">/form-control-infrastructure.html</a>  ( <a href="https://whatpr.org/html/10633/2e87c0f...bcd9669/form-control-infrastructure.html" title="Last updated on May 5, 2025, 2:02 PM UTC (bcd9669)">diff</a> )
<a href="https://whatpr.org/html/10633/form-elements.html" title="Last updated on May 5, 2025, 2:02 PM UTC (bcd9669)">/form-elements.html</a>  ( <a href="https://whatpr.org/html/10633/2e87c0f...bcd9669/form-elements.html" title="Last updated on May 5, 2025, 2:02 PM UTC (bcd9669)">diff</a> )
<a href="https://whatpr.org/html/10633/index.html" title="Last updated on May 5, 2025, 2:02 PM UTC (bcd9669)">/index.html</a>  ( <a href="https://whatpr.org/html/10633/2e87c0f...bcd9669/index.html" title="Last updated on May 5, 2025, 2:02 PM UTC (bcd9669)">diff</a> )
<a href="https://whatpr.org/html/10633/indices.html" title="Last updated on May 5, 2025, 2:02 PM UTC (bcd9669)">/indices.html</a>  ( <a href="https://whatpr.org/html/10633/2e87c0f...bcd9669/indices.html" title="Last updated on May 5, 2025, 2:02 PM UTC (bcd9669)">diff</a> )
<a href="https://whatpr.org/html/10633/infrastructure.html" title="Last updated on May 5, 2025, 2:02 PM UTC (bcd9669)">/infrastructure.html</a>  ( <a href="https://whatpr.org/html/10633/2e87c0f...bcd9669/infrastructure.html" title="Last updated on May 5, 2025, 2:02 PM UTC (bcd9669)">diff</a> )
<a href="https://whatpr.org/html/10633/interactive-elements.html" title="Last updated on May 5, 2025, 2:02 PM UTC (bcd9669)">/interactive-elements.html</a>  ( <a href="https://whatpr.org/html/10633/2e87c0f...bcd9669/interactive-elements.html" title="Last updated on May 5, 2025, 2:02 PM UTC (bcd9669)">diff</a> )
<a href="https://whatpr.org/html/10633/parsing.html" title="Last updated on May 5, 2025, 2:02 PM UTC (bcd9669)">/parsing.html</a>  ( <a href="https://whatpr.org/html/10633/2e87c0f...bcd9669/parsing.html" title="Last updated on May 5, 2025, 2:02 PM UTC (bcd9669)">diff</a> )

annevk commented:
It seems this is missing a lot of the boilerplate that new elements normally have as well as changes to content models, indexes, etc. See also this checklist at the top of `source`:
```
 !   Adding a new element involves editing the following sections:
 !    - section for the element itself
 !    - descriptions of the element's categories
 !    - images/content-venn.svg
 !    - syntax, if it's void or otherwise special
 !    - parser, if it's not phrasing-level
 !    - rendering
 !    - obsolete section
 !    - element, attribute, content model, and interface indices
```

josepharhar commented:
Thanks!

- I added a dl with a bunch of metadata for the element itself.
- I didn't include this element in any categories
- I didn't update the content-venn image because I didn't add it to any categories.
- I didn't update syntax or parsing because it doesn't have any special rules and I didn't implement anything in the HTML parser for this.
- I didn't update rendering because I noticed that elements like `<div>` and `<span>` don't have sections in there, and selectedoption is supposed to render like those elements.
- I didn't update the obsolete section because nothing is becoming obsolete...?
- I updated the element and interface indexes

annevk commented:
This still seems incomplete. Where does the `button` element indicate this element can be a descendant, for instance? If you don't adjust the categories, you still need to account for how the categories are used.

annevk commented on source:
```diff
@@ -55816,6 +55882,145 @@ interface <dfn interface>HTMLLegendElement</dfn> : <span>HTMLElement</span> {
+    </ol>
+   </li>
+
+   <li><p>Run <span data-x="dom-parentnode-replacechildren">replaceChildren</span> on
+   <var>selectedOption</var> given <var>nodes</var>.</p></li>
```
You cannot invoke public API.

Also, per the discussion in #10520 this is not the cloning behavior that's implemented?

josepharhar replied:
> this is not the cloning behavior that's implemented?

It was this way before I [made it use a MutationObserver](https://chromium-review.googlesource.com/c/chromium/src/+/5758741). I am planning on [reverting that patch](https://chromium-review.googlesource.com/c/chromium/src/+/5909975) in order to support the synchronous timing which was able to get consensus at WHATNOT.

> You cannot invoke public API.

Good catch, thanks! I inlined the steps from the dom spec.

josepharhar commented:
> This still seems incomplete. Where does the `button` element indicate this element can be a descendant, for instance? If you don't adjust the categories, you still need to account for how the categories are used.

Thanks, I updated the content model of the button element

domfarolino commented on source:
```diff
@@ -55816,6 +55885,151 @@ interface <dfn interface>HTMLLegendElement</dfn> : <span>HTMLElement</span> {
+
+   <li><p>Run <span>clone an option into a selectedoption</span> given the first <code>option</code>
+   in <var>firstAncestorSelect</var>'s <span data-x="concept-select-option-list">list of
+   options</span> whose <span data-x="concept-option-selectedness">selectedness</span> is true and
```
```suggestion
   options</span> whose <span data-x="concept-option-selectedness">selectedness</span> is true, and
```

josepharhar replied:
done

domfarolino commented on source:
```diff
@@ -55816,6 +55885,151 @@ interface <dfn interface>HTMLLegendElement</dfn> : <span>HTMLElement</span> {
+   <li><p>Let <var>convertedNode</var> be the result of <span>convert nodes into a node</span> given
+   <var>nodes</var> and <var>option</var>'s <span>node document</span>.</p></li>
+
+   <li><p><span>Ensure pre-insertion validity</span> of <var>convertedNode</var> into
```
At this point, `<var>convertedNodes</var>` is expected to be null if `<var>option</var>` is null, right? I think that's the case because if `<var>option</var>` is null, then "convert nodes into a node" would just return null.

And if `<var>convertedNodes</var>` is null here, then I think we would always fail the pre-insertion validity check, per its step 4. Basically I think we should just handle the "clear-the-selectedoption" case separately from the "actual clone" case, instead of merging them together and having the clone algorithm accept a null option.

josepharhar replied:
I created a separate algorithm for clearing called "clear a selectedoption"

domfarolino commented on source:
```diff
@@ -55816,6 +55885,151 @@ interface <dfn interface>HTMLLegendElement</dfn> : <span>HTMLElement</span> {
+  <var>selectedoption</var>, are:</p>
+
+  <ol>
+   <li><p>Let <var>firstAncestorSelect</var> be null.</p></li>
```
This is an ultra nit but I think I would rename this `nearestSelectAncestor`. The ordering of `ancestorSelect` feels slightly odd, so I'd suggest `selectAncestor`, and "nearest" is less ambiguous than "first" (since first is not directional, regarding the list of all ancestors).

josepharhar replied:
done

domfarolino commented on source:
```diff
@@ -55816,6 +55885,151 @@ interface <dfn interface>HTMLLegendElement</dfn> : <span>HTMLElement</span> {
+  support <code>selectedoption</code> elements:</p>
+
+  <ol>
+   <li><p>Run <span>maybe clone option into selectedoption</span> given <var>element</var>.</p></li>
```
Is this necessary if we are already modifying the global insertion and removal steps too?

josepharhar replied:
I'm planning on removing modifications to the insertion/removal steps shortly, I'll come back to this when that is done

josepharhar replied:
Ah right this is for attribute modification - I removed this.

scottaohara commented on source:
```diff
@@ -52991,6 +53000,8 @@ You cannot submit this form when the field is incorrect.</samp></pre>
    <dd><span>Phrasing content</span>, but there must be no <span>interactive content</span>
    descendant and no descendant with the <code data-x="attr-tabindex">tabindex</code> attribute
    specified.</dd>
+   <dd>If the element is the first child of a <code>select</code> element, then it must have one
```
@josepharhar @mfreed7, this 'must' is in opposition to what we were talking about during the OpenUI call yesterday - which is part of the reason why i wasn't quite sure how to respond when a use case was mentioned where a button wouldn't have a selectedoption element within it.

mfreed7 replied:
I would 100% agree, the `<selectedoption>` shouldn't be required at all. @josepharhar I'm not sure if the "must" was intentional - seems like no. But if so, let us know why.

scottaohara replied:
i think this comes from our original talks about how to construct the content model, so unless i'm not remembering correctly, this likely came from what I originally wrote.  But, as things have evolved with this element/its use cases, this just needs to be updated to match.

josepharhar replied:
I changed it to zero or one selectedoption

josepharhar commented on source:
```diff
@@ -55816,6 +55885,151 @@ interface <dfn interface>HTMLLegendElement</dfn> : <span>HTMLElement</span> {
+   <li><p>Let <var>nodes</var> be « ».</p></li>
+
+   <li>
+    <p>If <var>option</var> is not null, then for each <var>child</var> of <var>option</var>'s <span
```
Thanks! I created a separate algorithm for clearing the selectedoption

josepharhar commented on source:
```diff
@@ -55816,6 +55885,151 @@ interface <dfn interface>HTMLLegendElement</dfn> : <span>HTMLElement</span> {
+
+  <p>The <code>selectedoption</code> element reflects the contents of a <code>select</code>
+  element's currently selected <code>option</code> element. <code>selectedoption</code> elements can
+  be used to declaratively show the selected <code>option</code> element's contents in a different
```
Yeah this only makes sense if we do the selectedoptionelement attribute which I am postponing, so I reworded it to remove this.

josepharhar commented:
I just made some changes:
- Renamed from `<selectedoption>` to `<selectedcontent>` based on the resolution in https://github.com/openui/open-ui/issues/1112
- Stopped observing mutations within the selected `<option>` and replaced with `<option>` end tag parser cloning based on WHATNOT conversation and discussions in https://github.com/whatwg/html/issues/10520 and resolution in https://github.com/openui/open-ui/issues/825#issuecomment-2436124668

domfarolino commented on source:
```diff
@@ -54363,6 +54371,51 @@ interface <dfn interface>HTMLOptionElement</dfn> : <span>HTMLElement</span> {
+  <ol>
+   <li>
+    <p>For each <var>ancestor</var> of <var>option</var>'s <span
+    data-x="ancestor">ancestors</span>:</p>
```
I think I would also express the ordering in the name of this algorithm. Maybe something like:

```
To get the <dfn>option element's nearest select ancestor</dfn>
```

And I think I would markup `option` and `select` personally (similar to what Anne suggested further above regarding `selectedcontent`).

Also related: https://github.com/whatwg/html/pull/10633#discussion_r1821031206 (regarding ordering of "select" and "ancestor", which is a nit I don't feel that strongly about).

annevk replied:
I don't think that's sufficient to define ordering for the iteration though. It needs to be part of the iteration somehow. cc @zcorpan

josepharhar replied:
As I mentioned in at least one of the other PRs, I don't know what would be the best way to do this. I have seen other text add "in tree order" to the end of stuff like this to indicate ordering, but I think that "in tree order" might be the wrong order because I want to start at the parent and then walk up the tree.

I could add "starting at option's parent node"

domfarolino replied:
Yeah that's why I said "also"

josepharhar replied:
Or I could replace the "for each" loop with a while loop that manually updates the variable like this: https://html.spec.whatwg.org/multipage/interactive-elements.html#ancestor-details-revealing-algorithm

annevk replied:
You could say "in reverse tree order".

josepharhar replied:
done!

domfarolino commented on source:
```diff
@@ -56023,6 +56076,156 @@ interface <dfn interface>HTMLLegendElement</dfn> : <span>HTMLElement</span> {
+
+   <li>
+    <p>For each <var>child</var> of <var>option</var>'s <span
+    data-x="concept-tree-child">children</span>:</p>
```
Hmm, I don't know? The children member is an ordered set, and I think we can just iterate over ordered sets normally, right? Just like the `children` DOM getter doesn't specify the order of the collection returned: https://dom.spec.whatwg.org/#dom-parentnode-children.

annevk replied:
Oh right. I forgot I made that change at some point and was thinking about an earlier definition of child.

josepharhar replied:
Cool, I'm guessing this is ok as-is then

domfarolino commented on source:
```diff
@@ -56023,6 +56076,156 @@ interface <dfn interface>HTMLLegendElement</dfn> : <span>HTMLElement</span> {
+    </ol>
+   </li>
+
+   <li><p>Let <var>convertedNode</var> be the result of <span>convert nodes into a node</span> given
```
nit: this feels like it should be `the result of running` or `the result of <span> converting`

josepharhar replied:
I did "the result of running"

domfarolino commented on source:
```diff
@@ -56023,6 +56076,156 @@ interface <dfn interface>HTMLLegendElement</dfn> : <span>HTMLElement</span> {
+    </ol>
+   </li>
+
+   <li><p>Let <var>convertedNode</var> be the result of <span>convert nodes into a node</span> given
```
Maybe for readability/clarity this variable could be `documentFragment`?

domfarolino commented on source:
```diff
@@ -56023,6 +56076,156 @@ interface <dfn interface>HTMLLegendElement</dfn> : <span>HTMLElement</span> {
+   <var>nodes</var> and <var>option</var>'s <span>node document</span>.</p></li>
+
+   <li><p><span>Ensure pre-insertion validity</span> of <var>convertedNode</var> into
+   <var>SelectedContent</var> before null.</p></li>
```
```suggestion
   <var>selectedcontent</var> before null.</p></li>
```

domfarolino commented on source:
```diff
@@ -56023,6 +56076,156 @@ interface <dfn interface>HTMLLegendElement</dfn> : <span>HTMLElement</span> {
+  <ol>
+   <li><p><span data-x="concept-node-remove-ext">Remove</span> all <span
+   data-x="concept-tree-child">children</span> of <var>selectedcontent</var>, in <span>tree
+   order</span></p></li>
```
```suggestion
   order</span>.</p></li>
```

josepharhar replied:
done

domfarolino commented on source:
```diff
@@ -56023,6 +56076,156 @@ interface <dfn interface>HTMLLegendElement</dfn> : <span>HTMLElement</span> {
+
+   <li><p><span data-x="list append">Append</span> <var>selectedcontent</var> to
+   <var>nearestSelectAncestor</var>'s <span>select descendant selectedcontent
+   elements</span>.</p></li>
```
I guess this is related to Anne's question above, but do we support multiple `selectedcontent` elements or no?

domfarolino commented on source:
```diff
@@ -56023,6 +56076,156 @@ interface <dfn interface>HTMLLegendElement</dfn> : <span>HTMLElement</span> {
+   <var>insertedNode</var>.</p></li>
+
+   <li><p>Otherwise, run <span>clone an option into a selectedcontent</span> given <var>option</var>
+   and <var>insertedNode</var>.</p></li>
```
Hmm, this scares me a bit! You're calling the "clone an option" algorithm from inside the `selectedcontent`'s insertion steps, but [these steps must not run script](https://dom.spec.whatwg.org/#concept-node-insert-ext), yet the "clone an option" algorithm _can_ run script, IIUC, because of the "replace all" call.

Given that, I'm actually thinking we should run all of these steps not inside the selected content insertion steps, but maybe its [post-connection steps](https://dom.spec.whatwg.org/#concept-node-post-connection-ext) instead — those can run script. A test for this would be great, to confirm my suspicion. It might require some tricks to clone a script that will run during the "clone an option" algorithm.

josepharhar replied:
Sure I'm happy to do this. Do you know how post-connection steps are implemented in chromium?

domfarolino replied:
via [`DidNotifySubtreeInsertionsToDocument()`](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/dom/node.h;l=850;drc=d6a51e45b8900368202f33b04585980711e629ba)

josepharhar replied:
Cool, I used post-insertion steps instead

domfarolino reviewed (commented):
Nothing major except I'm concerned about using selectedcontent element's insertion steps instead of post-connection steps. This is because I believe the steps as currently specified can run script, and script must not run during an element's insertion steps.

josepharhar commented on source:
```diff
@@ -53198,6 +53201,8 @@ You cannot submit this form when the field is incorrect.</samp></pre>
    descendant and no descendant with the <code data-x="attr-tabindex">tabindex</code> attribute
    specified.</dd>
+   <dd>If the element is the first child of a <code>select</code> element, then it may have zero or
+   one <code>selectedcontent</code> element.</dd>
```
Yes, the previous line about phrasing content also applies so that you can have phrasing content and a selectedcontent element. I added the word "also" to make this more clear. I could also combine this with the previous `<dd>`, what do you think?

annevk replied:
I think in element definitions thus far the `dd` elements are mutually exclusive. Not additive.

josepharhar replied:
Ok, I combined it with the above dd

josepharhar commented on source:
```diff
@@ -53511,6 +53516,9 @@ interface <dfn interface>HTMLSelectElement</dfn> : <span>HTMLElement</span> {
   element, and all the <code>option</code> element children of all the <code>optgroup</code> element
   children of the <code>select</code> element, in <span>tree order</span>.</p>
 
+  <p>Every <code>select</code> element has <dfn>select descendant selectedcontent elements</dfn>,
```
Ok, I changed it to `descendant <code>selectedcontent</code> elements`.

josepharhar commented on source:
```diff
@@ -54363,6 +54371,51 @@ interface <dfn interface>HTMLOptionElement</dfn> : <span>HTMLElement</span> {
+
+  <p>Whenever an <code>option</code> <var>option</var>'s <span
+  data-x="concept-option-selectedness">selectedness</span> is set to true, run <span>maybe clone an
+  option into selectedcontent</span> given <var>option</var>.</p>
```
I moved this to the "send select update notifications" algorithm

josepharhar commented on source:
```diff
@@ -54363,6 +54371,51 @@ interface <dfn interface>HTMLOptionElement</dfn> : <span>HTMLElement</span> {
+    <p>then for each <var>selectedcontent</var> of <var>option</var>'s <span>option element
+    ancestor select</span>'s <span>select descendant selectedcontent elements</span>, run
+    <span>clone an option into a selectedcontent</span> given <var>option</var> and
+    <var>selectedcontent</var>.</p>
```
It isn't my goal to support multiple selectedcontent elements, and I don't know of a use case for multiple.

I like maintaining a list of descendant selectedcontent elements because it avoids the need to traverse through all the descendants of the select element to find another selectedcontent element in the case that a selectedcontent element is removed.

Now that I think about it, I'm probably OK with just performing a clone into the first selectedoption element in this list. I updated the text to take the first element from the list and added logic in some other places to make sure that only the first inserted selectedcontent element is kept up to date.

josepharhar commented on source:
```diff
@@ -56023,6 +56076,156 @@ interface <dfn interface>HTMLLegendElement</dfn> : <span>HTMLElement</span> {
+  element's currently selected <code>option</code> element. <code>selectedcontent</code> elements can
+  be used to declaratively show the selected <code>option</code> element's contents within the
+  <code>select</code>'s <code>button</code>, with alternate rendering based on different selectors
+  in the author's stylesheet.</p>
```
done

josepharhar commented on source:
```diff
@@ -56023,6 +56076,156 @@ interface <dfn interface>HTMLLegendElement</dfn> : <span>HTMLElement</span> {
+  <p>Every time the selected <code>option</code> of a <code>select</code> switches from one option
+  to another, the <code>selectedcontent</code> element removes all of its children and replaces them
+  with a new cloned copy of the DOM structure of the <code>select</code>'s selected
+  <code>option</code> element.</p>
```
I added the `w-dev` attribute to the `<p>`. How does that look?

josepharhar replied:
I'm guessing this looks OK

josepharhar commented:
I just added a "disabled" concept to help prevent infinite loops during cloning due to putting a selectedcontent element inside multiple select elements, another selectedcontent element, or an option element.

josepharhar commented:
I made the select.value and select.selectedIndex setters trigger cloning to follow this resolution: https://github.com/openui/open-ui/issues/1119#issuecomment-2450565150

dbaron commented on source:
```diff
@@ -56214,6 +56289,213 @@ interface <dfn interface>HTMLLegendElement</dfn> : <span>HTMLElement</span> {
+      <p>If <var>ancestor</var> is a <code>select</code> element, then:</p>
+
+      <ol>
+       <li><p>If <var>nearestAncestorSelect</var> is null, then set <var>nearestAncestorSelect</var>
+       to <var>select</var>.</p></li>
```
I think this `nearestAncestorSelect` should be `nearestSelectAncestor`, and `select` should be `ancestor`

josepharhar replied:
done, thanks!

dbaron commented on source:
```diff
@@ -56214,6 +56289,213 @@ interface <dfn interface>HTMLLegendElement</dfn> : <span>HTMLElement</span> {
+   <li><p>If <var>selectedcontent</var> is null, then return.</p></li>
+
+   <li><p>Let <var>option</var> be the first <code>option</code> in
+   <var>nearestSelectAncestor</var>'s <span data-x="concept-select-option-list">list of
```
I think this `nearestSelectAncestor` should be `select`

josepharhar replied:
done, thanks!

dbaron reviewed (commented):
A few comments from looking at parts of this:

josepharhar commented:
In response to feedback from @dbaron, I made the one selectedcontent element which receives updates be the first one in tree order instead of the first one to be inserted. I also removed the select descendant selectedoption elements list.

hsivonen commented:
Adding a proper backlink to: https://github.com/mozilla/standards-positions/issues/1142

I'm worried about breaking the implied invariant that the parser creates exactly one element node per (possibly implied) non-erroneous start tag. I don't at this time have a concrete end-to-end scenario of what badness exactly may ensue, but it's a bad sign about breaking an invariant leading to issues that then need addressing that there's already a remark about preventing infinite loops upthread and a separate issue about the timing of cloning (#10520).

In the error handling case where one element per start tag does not hold, the spec says "[Create an element for the token](https://html.spec.whatwg.org/#create-an-element-for-the-token) for which the element node was created, in the [HTML namespace](https://infra.spec.whatwg.org/#html-namespace)" instead of cloning a formatting element. This is in response to https://www.w3.org/Bugs/Public/show_bug.cgi?id=6743 , which was motivated by avoidance of data flows from the live DOM into the tree builder algorithm. However, it looks like the `option` cloning as proposed could happen on the main thread in a way that does not affect the subsequent decisions of the tree builder algorithm. 

Do I understand correctly that cloning into the regular DOM instead of a UA shadow DOM came from https://github.com/w3c/csswg-drafts/issues/10242#issuecomment-2250667040 ?

josepharhar commented:
> Do I understand correctly that cloning into the regular DOM instead of a UA shadow DOM came from https://github.com/w3c/csswg-drafts/issues/10242#issuecomment-2250667040 ?

Yes. We had many lengthy discussions about this, since cloning into shadow DOM avoids a few problems. But it also introduced a myriad of issues that were much more serious than the ones posed by cloning into regular DOM.

domfarolino commented on source:
```diff
@@ -54090,14 +54106,23 @@ interface <dfn interface>HTMLSelectElement</dfn> : <span>HTMLElement</span> {
+   options</span> to false.</p></li>
+
+   <li><p>The first <code>option</code> element in the <span
+   data-x="concept-select-option-list">list of options</span>, in <span>tree order</span>, whose
```
The list of options is already defined to be ordered in tree order, so I think this reference is redundant. I guess I'd probably remove it.

josepharhar replied:
removed

domfarolino commented on source:
```diff
@@ -54090,14 +54106,23 @@ interface <dfn interface>HTMLSelectElement</dfn> : <span>HTMLElement</span> {
+   data-x="concept-select-option-list">list of options</span>, in <span>tree order</span>, whose
+   <span data-x="concept-option-value">value</span> is equal to the given new value, if any, must
+   have its <span data-x="concept-option-selectedness">selectedness</span> set to true and its <span
+   data-x="concept-option-dirtiness">dirtiness</span> set to true.</p></li>
```
You're specifying that the first candidate option _must_ have these qualities, but you're not setting them yourself. Do you mean to? I would probably do this a little more algorithmically: just iterate through the list of options, filter out the one you want, and set the selectedness and dirtiness accordingly. WDYT?

josepharhar replied:
Agreed, done

domfarolino commented on source:
```diff
@@ -54090,14 +54106,23 @@ interface <dfn interface>HTMLSelectElement</dfn> : <span>HTMLElement</span> {
+
+   <li><p>The first <code>option</code> element in the <span
+   data-x="concept-select-option-list">list of options</span>, in <span>tree order</span>, whose
+   <span data-x="concept-option-value">value</span> is equal to the given new value, if any, must
```
Is it just a straight-up equal comparison we want? Chromium seems to [do some stuff with whitespace](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/html/forms/html_option_element.cc;l=252-253;drc=27d34700b83f381c62e3a348de2e6dfdc08364b8), do we need to do the same here?

josepharhar replied:
The chromium code there which does stuff with whitespace is accounted for in the option's value here (option value uses option text in the spec): https://html.spec.whatwg.org/#dom-option-text

Assigning to select.value doesn't seem to do anything special with whitespace: https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/html/forms/html_select_element.cc;l=636-671;drc=fc2575ac635633b2972b49eec663250ac05a1a53

domfarolino replied:
Ah OK got it. So I don't think anything is required here then. Can we resolve this?

josepharhar replied:
yes

domfarolino commented on source:
```diff
@@ -54554,6 +54579,52 @@ interface <dfn interface>HTMLOptionElement</dfn> : <span>HTMLElement</span> {
+   <li><p>Return null.</p></li>
+  </ol>
+
+  <p>To <dfn>maybe clone an option into selectedcontent</dfn>, given an <code>option</code>
```
I think I would link to `<code>option</code>` and `<code>selectedcontent</code>` here like we do in a lot of the other algorithm declarations.

josepharhar replied:
done

domfarolino commented on source:
```diff
@@ -54554,6 +54579,52 @@ interface <dfn interface>HTMLOptionElement</dfn> : <span>HTMLElement</span> {
+     <li><p><var>select</var> is not null;</p></li>
+
+     <li><p><var>select</var> does not have the <code data-x="attr-select-multiple">multiple</code>
+     attribute;</p></li>
```
Just FYI this condition is redundant, since it's also checked when we retrieve select's "enabled selectedcontent". Might be OK to leave it here for clarity, but could be simpler to remove. I'll let you decide; I just wanted to point it out in case you didn't realize.

josepharhar replied:
Thanks! I removed it since I wouldn't want to forget about it later if we decide to do something with selectedcontent in select multiple.

domfarolino commented on source:
```diff
@@ -56214,6 +56285,221 @@ interface <dfn interface>HTMLLegendElement</dfn> : <span>HTMLElement</span> {
+  <p>The <code>selectedcontent</code> element reflects the contents of a <code>select</code>
+  element's currently selected <code>option</code> element. <code>selectedcontent</code> elements
+  can be used to declaratively show the selected <code>option</code> element's contents within the
+  <code>select</code> element's <code>button</code> element.</p>
```
This button element isn't defined anywhere special right? Elsewhere we refer to it as the "first child button", so maybe we can clarify that here too? (Unless we have a dfn for it to reference it more directly that I'm not aware of)

josepharhar replied:
That's right, it isn't defined anywhere special. It is also referred to in the base appearance spec PR. I added "first child" before "button" to make it more clear.

domfarolino commented on source:
```diff
@@ -56214,6 +56285,221 @@ interface <dfn interface>HTMLLegendElement</dfn> : <span>HTMLElement</span> {
+
+  <p><code>selectedcontent</code> elements become associated with a <code>select</code> element when
+  the <code>selectedcontent</code> is a <span>descendant</span> of the <code>select</code>'s first
+  child <code>button</code>.</p>
```
But the "get a select's enabled selectedcontent" algorithm doesn't limit its search to the select element's first child button. Is this out of date?

josepharhar replied:
You're right, I pretty much just said this because it is meant to be used inside the button. I removed this paragraph.

domfarolino replied:
Great, this can be resolved right? (Just want to double check with you instead of resolving it myself)

josepharhar replied:
Yes

domfarolino commented on source:
```diff
@@ -56214,6 +56285,221 @@ interface <dfn interface>HTMLLegendElement</dfn> : <span>HTMLElement</span> {
+   <li><p>Let <var>documentFragment</var> be the result of running <span>convert nodes into a
+   node</span> given <var>nodes</var> and <var>option</var>'s <span>node document</span>.</p></li>
+
+   <li><p><span>Ensure pre-insertion validity</span> of <var>documentFragment</var> into
```
Now that I think about it, do we actually have to call this? Are there any circumstances where this could actually throw an exception? The only one that stands out to me is if `nodes` contains a doctype, but I don't think that's even possible, right? How does the Chromium impl look here? Are there any known cases where we see an exception?

domfarolino replied:
Yeah because an author can call `$0.replaceChildren($0.parentNode)` which would trip the second pre-insertion validity condition and throw an exception. I don't think we can run into that in our case, since we definitely aren't ever trying to insert parents into their children, right?

So the real question is about whether we need this call for _the other pre-insertion conditions_. I don't think we can ever trip those either though.

> We are also calling it within ReplaceChildren in chromium I think: https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/html/forms/html_selected_content_element.cc;l=35;drc=8c797f761029413678dc8d9c2308d78336046ab8

OK this is great to see. Note that we're asserting no exception in the impl, meaning we're guaranteeing that all pre-insertion checks will pass, and never rseult in an error. So I think we can remove this from the spec.

josepharhar replied:
removed, thanks!

domfarolino commented on source:
```diff
@@ -56214,6 +56285,221 @@ interface <dfn interface>HTMLLegendElement</dfn> : <span>HTMLElement</span> {
+   <li><p>Let <var>passedFirstSelectedcontent</var> be false.</p></li>
+
+   <li>
+    <p>For each <span>descendant</span> of <var>select</var> in <span>tree order</span>:</p>
```
I think you need to declare the variable `<var>descendant</var>` after the `<span>`. It should probably read: `For each <var>descendant</var> of <var>select</var>'s <span data-x="descendant">descendants</span>` to be consistent with how we access and iterate over ancestors in the post-connection steps algorithm.

josepharhar replied:
thanks, fixed

domfarolino commented on source:
```diff
@@ -56214,6 +56285,221 @@ interface <dfn interface>HTMLLegendElement</dfn> : <span>HTMLElement</span> {
+
+    <ol>
+     <li>
+      <p>If <var>descendant</var> is a <code>selectedcontent</code> element, then:</p>
```
nit: just for readability I might hoist this condition up to the for loop just to decrease the amount of indentations we have. It'd then read:

`For each <var>descendant</var> of <var>select</var>'s <span data-x="descendant">descendants</span> in <span>tree order</span> that is a <code>selectedcontent</code> element:`

josepharhar replied:
done

domfarolino commented on source:
```diff
@@ -56214,6 +56285,221 @@ interface <dfn interface>HTMLLegendElement</dfn> : <span>HTMLElement</span> {
+
+      <ol>
+       <li><p>If <var>nearestSelectAncestor</var> is null, then set <var>nearestSelectAncestor</var>
+       to <var>select</var>.</p></li>
```
So the Chromium implementation seems to [unconditionally assign](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/html/forms/html_selected_content_element.cc;l=66;drc=8c797f761029413678dc8d9c2308d78336046ab8) `<var>nearestSelectAncestor</var>` in this case, not limiting assignment to the non-null case. That seems to not matter in the implementation since the last two steps of this algorithm that actually use the nearest select ancestor don't appear the impl. What am I missing?

josepharhar replied:
The difference here is that chromium's select elements still keep a list of descendant selectedcontent elements as a performance optimization which I removed from this PR earlier, and all ancestor selects get this selectedcontent element added to their list of descendants, even when there are nested select elements.

The implementation doesn't seem to correctly remove them in the corresponding HTMLSelectedContentElement::RemovedFrom, so I'm going to change the implementation to match the spec

josepharhar replied:
https://chromium-review.googlesource.com/c/chromium/src/+/6196578

domfarolino replied:
Cool! I think this can be resolved now then, right?

josepharhar replied:
Yes

domfarolino commented on source:
```diff
@@ -54070,13 +54076,22 @@ interface <dfn interface>HTMLSelectElement</dfn> : <span>HTMLElement</span> {
+   <code>option</code> elements in the <span data-x="concept-select-option-list">list of
+   options</span> to false.</p></li>
+
+   <li><p>The <code>option</code> element in the <span data-x="concept-select-option-list">list of
```
I think this suffers from a very similar problem to https://github.com/whatwg/html/pull/10633#discussion_r1927682415, in that we're specifying requirements but not taking action.

josepharhar replied:
agreed and fixed, thanks!

domfarolino commented on source:
```diff
@@ -54554,6 +54588,49 @@ interface <dfn interface>HTMLOptionElement</dfn> : <span>HTMLElement</span> {
 
   </div>
 
+  <p>To get the <dfn><code>option</code> element nearest ancestor <code>select</code></dfn> given an
```
I think editorially in HTML we generally end all "To foo" algorithms, with ", run these steps". And if they return a value, then we say ", run these steps. They return a bar." See algorithms in https://html.spec.whatwg.org/#resolving-urls for example. So I'd have a slight editorial preference to changing all of the algorithms here to match that pattern.

josepharhar replied:
I fixed this and the algorithm below

domenic replied:
Sorry, this isn't quite right. Modern style doesn't include "run these steps" unless you also specify a return type: https://infra.spec.whatwg.org/#algorithm-declaration

josepharhar replied:
ok i removed "run these steps" from the declaration which doesn't mention a return type

domfarolino reviewed (approved):
OK I think this LGTM!

I would like some clarification in the OP as to whether #10520 is fully resolved and fixed by this PR. And we'll need to check the rest of the boxes in the OP and fill out the needed information.

annevk commented on source:
```diff
@@ -54070,13 +54076,33 @@ interface <dfn interface>HTMLSelectElement</dfn> : <span>HTMLElement</span> {
+   <li><p>Let <var>firstMatchingOption</var> be null.</p></li>
+
+   <li>
+    <p>For each <var>option</var> in the <code>select</code>'s <span
```
```suggestion
    <p>For each <var>option</var> of <span>this</span>'s <span
```

josepharhar replied:
done

annevk replied:
It's only partially done it seems. in -> of is also important.

josepharhar replied:
Thanks, my bad. Should be fixed now.

I wish the "commit suggestion" button actually worked for these PRs

annevk commented on source:
```diff
@@ -54070,13 +54076,33 @@ interface <dfn interface>HTMLSelectElement</dfn> : <span>HTMLElement</span> {
+     to false.</p></li>
+
+     <li><p>If <var>firstMatchingOption</var> is null and <var>option</var>'s <span
+     data-x="concept-option-index">index</span> is equal to the given new value, then set
```
```suggestion
     data-x="concept-option-index">index</span> is equal to the given value, then set
```

josepharhar replied:
done

annevk commented on source:
```diff
@@ -54070,13 +54076,33 @@ interface <dfn interface>HTMLSelectElement</dfn> : <span>HTMLElement</span> {
+   <var>firstMatchingOption</var>'s <span data-x="concept-option-dirtiness">dirtiness</span> to
+   true.</p></li>
+
+   <li><p>Run <span>update a <code>select</code>'s <code>selectedcontent</code></span> given this
+   <code>select</code>.</p></li>
```
```suggestion
   <li><p>Run <span>update a <code>select</code>'s <code>selectedcontent</code></span> given
   <span>this</span>.</p></li>
```

josepharhar replied:
done

annevk commented on source:
```diff
@@ -54090,14 +54116,33 @@ interface <dfn interface>HTMLSelectElement</dfn> : <span>HTMLElement</span> {
+   <li><p>Let <var>firstMatchingOption</var> be null.</p></li>
+
+   <li>
+    <p>For each <var>option</var> in the <code>select</code>'s <span
+    data-x="concept-select-option-list">list of options</span>:</p>
```
Ditto.

josepharhar replied:
done

annevk commented on source:
```diff
@@ -54090,14 +54116,33 @@ interface <dfn interface>HTMLSelectElement</dfn> : <span>HTMLElement</span> {
+   <var>firstMatchingOption</var>'s <span data-x="concept-option-dirtiness">dirtiness</span> to
+   true.</p></li>
+
+   <li><p>Run <span>update a <code>select</code>'s <code>selectedcontent</code></span> given this
+   <code>select</code>.</p></li>
```
Ditto.

josepharhar replied:
done

annevk commented on source:
```diff
@@ -54090,14 +54116,33 @@ interface <dfn interface>HTMLSelectElement</dfn> : <span>HTMLElement</span> {
+     to false.</p></li>
+
+     <li><p>If <var>firstMatchingOption</var> is null and <var>option</var>'s <span
+     data-x="concept-option-value">value</span> is equal to the given new value, then set
```
Ditto.

josepharhar replied:
done

domfarolino commented on source:
```diff
@@ -132170,9 +132463,18 @@ document.body.appendChild(text);
+    <ol>
+     <li><p>Let <var>option</var> be the <span>current node</span>.</p></li>
+
+     <li><p>Pop <var>option</var> from the <span>stack of open elements</span>.</p></li>
```
Sorry for the nit, but how about we link to https://infra.spec.whatwg.org/#stack-pop here? We already include that dfn in the refs somewhere so it should be a tiny change.

domfarolino reviewed (approved):
This still LGTM; there are two things I'd like to see:
 - @josepharhar: could you address https://github.com/whatwg/html/pull/10633#pullrequestreview-2579350724 by adding some detail to the commit message in the OP, especially around the cloning strategy we ultimately went with and how it relates to the discussion in https://github.com/whatwg/html/issues/10520?
 - @scottaohara: how do you feel about https://github.com/w3c/html-aria/pull/528 and https://github.com/w3c/aria/pull/2344? Do you think those are ready to land at this point, and are just waiting on this PR?

Once I get some clarity on these two I think we're ready to land this; I will hold off until the end of the week.

josepharhar commented:
> * @josepharhar: could you address [Define the `<selectedcontent>` element #10633 (review)](https://github.com/whatwg/html/pull/10633#pullrequestreview-2579350724) by adding some detail to the commit message in the OP, especially around the cloning strategy we ultimately went with and how it relates to the discussion in [Timing of cloning for the `<selectedoption>` element #10520](https://github.com/whatwg/html/issues/10520)?

Done. Copying into this comment too in case the commit message gets clobbered by the thing that updates the commit message with the build of the PR: 

Some details on the cloning behavior, which were decided on in order to always make sure that the selectedcontent element is up to date and in order to reduce developer confusion:
- Cloning is done with an algorithm similar to the Node.cloneNode() function.
- Cloning is done synchronously at these times:
  - When an `</option>` end tag is parsed during HTML parsing.
  - A selectedcontent element is inserted, during its post-connection steps.
  - Immediately before input and change events are fired on a select element.
  - When the select.selectedIndex IDL is assigned to.

annevk commented on source:
```diff
@@ -54090,14 +54116,33 @@ interface <dfn interface>HTMLSelectElement</dfn> : <span>HTMLElement</span> {
+   <li><p>Let <var>firstMatchingOption</var> be null.</p></li>
+
+   <li>
+    <p>For each <var>option</var> in <span>this</span>'s <span
```
```suggestion
    <p>For each <var>option</var> of <span>this</span>'s <span
```

josepharhar replied:
fixed

annevk commented on source:
```diff
@@ -54554,6 +54599,50 @@ interface <dfn interface>HTMLOptionElement</dfn> : <span>HTMLElement</span> {
+
+  <ol>
+   <li>
+    <p>For each <var>ancestor</var> of <var>option</var>'s <span data-x="ancestor">ancestors</span>
+    in reverse <span>tree order</span>:</p>
```
```suggestion
    <p>For each <var>ancestor</var> of <var>option</var>'s <span data-x="ancestor">ancestors</span>,
    in reverse <span>tree order</span>:</p>
```

josepharhar replied:
done

annevk commented on source:
```diff
@@ -54554,6 +54599,50 @@ interface <dfn interface>HTMLOptionElement</dfn> : <span>HTMLElement</span> {
+     <li><p><var>select</var> is not null;</p></li>
+
+     <li><p><var>option</var>'s <span data-x="concept-option-selectedness">selectedness</span> is
+     true;</p></li>
```
```suggestion
     true; and</p></li>
```

annevk commented on source:
```diff
@@ -56214,6 +56303,210 @@ interface <dfn interface>HTMLLegendElement</dfn> : <span>HTMLElement</span> {
+
+  <p w-dev>Every time the selected <code>option</code> of a <code>select</code> switches from one
+  option to another, the <code>selectedcontent</code> element removes all of its children and
+  replaces them with a new cloned copy of the DOM structure of the <code>select</code>'s selected
```
"cloned copy" is a pleonasm.

josepharhar replied:
Ok, I removed "cloned". How does it look now?

annevk commented on source:
```diff
@@ -56214,6 +56303,210 @@ interface <dfn interface>HTMLLegendElement</dfn> : <span>HTMLElement</span> {
+
+    <ol>
+     <li><p>Let <var>childClone</var> be the result of running <span
+     data-x="concept-node-clone">clone</span> given <var>child</var>, null, true.</p></li>
```
and true?

But also, I don't think this works anymore now that concept-node-clone has named arguments?

josepharhar replied:
Thanks, I changed it to use named arguments. How does it look?

annevk commented on source:
```diff
@@ -56214,6 +56303,210 @@ interface <dfn interface>HTMLLegendElement</dfn> : <span>HTMLElement</span> {
+
+   <li>
+    <p>For each <var>ancestor</var> of <var>selectedcontent</var>'s <span
+    data-x="ancestor">ancestors</span> in reverse <span>tree order</span>:</p>
```
```suggestion
    data-x="ancestor">ancestors</span>, in reverse <span>tree order</span>:</p>
```

josepharhar replied:
done

annevk commented on source:
```diff
@@ -56214,6 +56303,210 @@ interface <dfn interface>HTMLLegendElement</dfn> : <span>HTMLElement</span> {
+
+    <ol>
+     <li>
+      <p>If <var>ancestor</var> is a <code>select</code> element, then:</p>
```
```suggestion
      <p>If <var>ancestor</var> is a <code>select</code> element:</p>
```

josepharhar replied:
done

annevk commented on source:
```diff
@@ -56214,6 +56303,210 @@ interface <dfn interface>HTMLLegendElement</dfn> : <span>HTMLElement</span> {
+  <ol>
+   <li>
+    <p>For each <var>ancestor</var> of <var>selectedcontent</var>'s <span
+    data-x="ancestor">ancestors</span> in reverse <span>tree order</span>:</p>
```
```suggestion
    data-x="ancestor">ancestors</span>, in reverse <span>tree order</span>:</p>
```

josepharhar replied:
done

annevk commented on source:
```diff
@@ -56214,6 +56303,210 @@ interface <dfn interface>HTMLLegendElement</dfn> : <span>HTMLElement</span> {
+
+   <li>
+    <p>For each <var>ancestor</var> of <var>oldParent</var>'s <span data-x="inclusive
+    ancestor">inclusive ancestors</span> in reverse <span>tree order</span>:</p>
```
```suggestion
    ancestor">inclusive ancestors</span>, in reverse <span>tree order</span>:</p>
```

josepharhar replied:
done

annevk commented on source:
```diff
@@ -132170,9 +132463,19 @@ document.body.appendChild(text);
-    <p>If the <span>current node</span> is an <code>option</code> element, then pop that node from
-    the <span>stack of open elements</span>. Otherwise, this is a <span>parse error</span>; ignore
-    the token.</p>
+    <p>If the <span>current node</span> is an <code>option</code> element, then:</p>
```
```suggestion
    <p>If the <span>current node</span> is an <code>option</code> element:</p>
```

scottaohara commented:
@domfarolino sorry, I missed your mention of me since I've already been mentioned in this thread, github doesn't make it clear when someone has been re-mentioned.

the html aam (2344) PR is waiting on this to merge.  It hasn't had any updates since I made some wording changes to a note in February.  There are various other HTML AAM PRs related to customizable select which are also ready to go once everything is ready to merge into the HTML spec.

the aria in html pr (528) is essentially ready to go as well.  No more spec updates needed, just some issues to file against conformance checkers.

annevk commented on source:
```diff
@@ -56601,6 +56690,211 @@ interface <dfn interface>HTMLLegendElement</dfn> : <span>HTMLElement</span> {
+   </li>
+
+   <li><p>Let <var>documentFragment</var> be the result of running <span>convert nodes into a
+   node</span> given <var>nodes</var> and <var>option</var>'s <span>node document</span>.</p></li>
```
This seems like the wrong primitive to use. Why not start out with a `DocumentFragment` and append each `childClone` to it as you go?

josepharhar replied:
Agreed, done

annevk commented on source:
```diff
@@ -56601,6 +56690,211 @@ interface <dfn interface>HTMLLegendElement</dfn> : <span>HTMLElement</span> {
+   <li><p><span data-x="concept-node-remove-ext">Remove</span> all <span
+   data-x="concept-tree-child">children</span> of <var>selectedcontent</var>, in <span>tree
+   order</span>.</p></li>
+  </ol>
```
I think we want to use replace all with nothing here, no? It would be good to write some mutation observer tests for this.

josepharhar replied:
Thanks, I changed it to replace all with null. I am working on mutation observer tests.

josepharhar replied:
Tests are being added here: https://chromium-review.googlesource.com/c/chromium/src/+/6471702

scottaohara commented:
@josepharhar not sure if this is the PR to comment on or not, but per our chat discussion it's probably worth calling out that while `<option label=foo>` can be used with the customizable select, and will render a visible label for the option in the select's popup - if selected nothing will be returned to the `selectedcontent` element - since that works by cloning the descendants of the option, and in this care there are none.

Someone can still benefit from broader styling capabilities _and_ still define option element label text using the attribute _if_ their customized select does not contain a button part (and thus no selectedcontent element).  Maybe it makes sense to call this out?   that way, if someone does attempt to use label attributes when defining their options, it's made clear "if you want to do this, then you shouldn't use the button part/selectedconent elements.  but, to have greater control, use child content of the option to define your labels, which will also allow for more styling capabilities for represented the selected content in the select's collapsed state.

or you know... word smith as you see fit

josepharhar commented:
Thanks Scott! I added a note here: https://github.com/whatwg/html/pull/10633/commits/bcd9669f81aa55139c9c370ab29482a953d89f1a

estelle commented:
Per https://github.com/mdn/content/pull/39411; "Dynamic modifications to the selected `<option>` element's content made after the `<select>` element has been created are not automatically cloned to the `<selectedcontent>` element, and must be manually updated by the developer."

This has usability and security risks. If the option can be updated without the content of the `<selectedcontent>` updating along with it without JS, this has the likelihood of leading to broken forms. It's also a security risk as the user may submit a form with an unintended value that is not visible due to malicious code. 

We don't want to add anything to the web platform that may potentially break things, make a form control less accessible, or add any vulnerabilities.

mfreed7 commented:
> If the option can be updated without the content of the `<selectedcontent>` updating along with it without JS, this has the likelihood of leading to broken forms.

I don't think this is possible without JS. To get this behavior, you *need* JS manipulating the DOM in a very particular way: the `<option>` contents must be mutated after that `<option>` is appended to the `<select>`. If the site is using plain html and css, then `<selectedcontent>` should "just work" in all cases.

> It's also a security risk as the user may submit a form with an unintended value that is not visible due to malicious code.

If the malicious code has DOM access, isn't this already true?

```
<select id=purchase_quantity>
  <option value=1000000> 1 </option>
</select>
```

This would submit the value "1000000" even though the user sees "1".

estelle commented:
>  If the site is using plain html and css, then `<selectedcontent>` should "just work" in all cases.

If the site is using JS, will it ever break the `<selectedcontent>`? My understanding is that is can, and does. 

```html
<select id=purchase_quantity>
  <option value=1000000> 1 </option>
</select>
```

while the above is currently and has always been possible, if the following is ever possible, the implementation needs to be fixed:

```html
 <select id=purchase_quantity>
 <button>
 <selectedcontent>1</selectedcontent>
 </button>
<option selected> 10</option>
</select>
```

This being pseudo-markup: Where "selected" is not necessarily an attribute, but rather the current state

titoBouzout commented:
An element that modifies the page by adding/removing elements seems like a bad idea. 

It will break any code that expects the structure of the html to be the same as it was served.
